### PR TITLE
Keychorn K7 Optical initial bringup

### PR DESCRIPTION
**Pull request by stdvar**
## Description

Add Keychron K7 optical support.
This requires changes in common matrix scan:
1) add delays in matrix scan needed for HW to switch
2) move matrix scan to extra row
3) add options to skip matrix scans to increase LED brightness (due to delays needed in the scan)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: drivers/sn32/matrix_sn32f248b.c [hunk 1]
```diff
@@ -36,6 +36,14 @@ static const pin_t row_pins[MATRIX_ROWS] = MATRIX_ROW_PINS;
...
 static const pin_t led_row_pins[LED_MATRIX_ROWS_HW] = LED_MATRIX_ROW_PINS;
 static const pin_t led_col_pins[LED_MATRIX_COLS] = LED_MATRIX_COL_PINS;
+#if defined(OPTICAL_MATRIX)
+//skipping some matrix scans will increase led brightness for optical matrix due to delays in the scan
+//500 gives noticeable flickering. should not set above 250
+static const uint8_t SKIP_SCANS = 30;
+#else
+static const uint8_t SKIP_SCANS = 0;
+#endif
+static uint8_t scan_count = 0;
```
Comment (CalcProgrammer1) [inline drivers/sn32/matrix_sn32f248b.c]: To save RAM where this is unused, I would make another define like SKIP_ENABLE.  If SKIP_ENABLE is not defined, leave matrix code completely unchanged.  If SKIP_ENABLE is defined, then compile in the skip code.  You can have it so OPTICAL_MATRIX automatically defines SKIP_ENABLE.  The same could be true for delay - use a define like DELAY_ENABLE which is automatically defined if OPTICAL_MATRIX is defined.  This way non-optical keyboards can use the delay too (as it seemed to be needed at least on some 248 non-B, could be needed on some 248B).

File: drivers/sn32/matrix_sn32f248b.c [hunk 2]
```diff
@@ -350,15 +369,18 @@ OSAL_IRQ_HANDLER(SN32_CT16B1_HANDLER) {
 
     SN_CT16B1->IC = mskCT16_MR24IC; // Clear match interrupt status
 
-    // Turn the selected row off
-    writePinLow(led_row_pins[current_row]);
-
-    // Turn the next row on
-    current_row = (current_row + 1) % LED_MATRIX_ROWS_HW;
+    // Turn row off
+    if (current_row < LED_MATRIX_ROWS_HW){
```
Comment (CalcProgrammer1) [inline drivers/sn32/matrix_sn32f248b.c]: Rework so behavior is unchanged if SKIP_ENABLE isn't defined

**Comment by dexter93** [comment]
`keychron/k7/rgb` and `keychron/k7/rgb/optical`

How did the linter not freak out about this? We should probably discuss on the dir structure for this

File: drivers/sn32/matrix_sn32f248b.c (hunk 1 above)
Comment (stdvar) [reply drivers/sn32/matrix_sn32f248b.c]: addressed

File: drivers/sn32/matrix_sn32f248b.c (hunk 2 above)
Comment (stdvar) [reply drivers/sn32/matrix_sn32f248b.c]: updated

**Comment by stdvar** [comment]
> `keychron/k7/rgb` and `keychron/k7/rgb/optical`
> 
> How did the linter not freak out about this? We should probably discuss on the dir structure for this

it did complain initially about absence of readme in that directory. so I added it there. =) lmk if you want to change structure.

File: drivers/sn32/matrix_sn32f24xx.c [hunk 3]
```diff
@@ -35,6 +35,17 @@ Ported to QMK by Stephen Peery <https://github.com/smp4488/>
 #    define PRESSED_KEY_PIN_STATE 0
 #endif
 
+#if defined(DELAY_ENABLE)
+void delay(void){
+    //should give 3500/48000000Mhz = 73us delay
+    //we want 73 micro thus to get value
+    for (int i = 0; i < 3500; ++i) {
```
Comment (dexter93) [inline drivers/sn32/matrix_sn32f24xx.c]: please make this value adjustable with a define

File: drivers/sn32/matrix_sn32f24xx.c [hunk 4]
```diff
@@ -157,6 +174,11 @@ void matrix_scan_keys(matrix_row_t raw_matrix[], uint8_t current_row){
                     // Disable the row
                     writePinHigh(row_pins[row_index]);
                 }
+
+                // Set all column pins input low to activate leds
+                for (uint8_t col_index = 0; col_index < MATRIX_COLS; col_index++) {
```
Comment (dexter93) [inline drivers/sn32/matrix_sn32f24xx.c:179]: what effect does this have on mechanical switches?

File: keyboards/keychron/k7/rgb/config.h [hunk 5]
```diff
@@ -39,7 +39,7 @@
 #define MATRIX_ROW_PINS { D11, D10, D9, D8, D7 }
 
 // Connects each switch in the dip switch to the GPIO pin of the MCU
-//#define DIP_SWITCH_PINS { D5, D6 }
+#define DIP_SWITCH_PINS { D4 }
```
Comment (dexter93) [inline keyboards/keychron/k7/rgb/config.h:42]: is there really one switch there?

**Comment by dexter93** [comment]
please fix the typo in the title

**Comment by stdvar** [comment]
updated per comments.

**Pull request merged**